scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/definition/action/CommonModelDataInfoProvider.scala

```diff
@@ -27,17 +31,57 @@ abstract class CommonModelDataInfoProvider(modelData: ModelData) {
     Seq.empty,
     ProductionServiceInvocationCollector,
     ComponentUseCase.TestDataGeneration,
+    NodesDeploymentData.empty,
     nonServicesLazyParamStrategy = LazyParameterCreationStrategy.default
   )
 
-  protected def compileSourceNode(
+  def collectAllSources(scenario: CanonicalProcess): List[SourceNodeData] = {
+    scenario.collectAllNodes.flatMap(asSource) ++ scenario.collectAllNodes.flatMap(asFragmentInputDefinition)
+  }
+
+  def compileSourceNode(
       source: SourceNodeData
   )(implicit jobData: JobData, nodeId: NodeId): ValidatedNel[ProcessCompilationError, Source] = {
     nodeCompiler.compileSource(source).compiledObject
   }
 
-  protected def collectAllSources(scenario: CanonicalProcess): List[SourceNodeData] = {
-    scenario.collectAllNodes.flatMap(asSource) ++ scenario.collectAllNodes.flatMap(asFragmentInputDefinition)
+  def collectAndCompileAllSourcesAndServices(
+      scenario: CanonicalProcess
+  )(implicit jobData: JobData): Map[NodeComponentInfo, Any] = {
+    compileSourcesAndService(scenario.collectAllNodes) ++
+      compileSourcesAndService(scenario.collectAllNodes.flatMap(asFragmentInputDefinition))
   }
 
+  private def compileSourcesAndService(nodes: List[NodeData])(implicit jobData: JobData): Map[NodeComponentInfo, Any] =
```

Comment:
I think, that we should compile all non-built-in nodes. Consider situations:
- custom node - it is more general case as service. it can be also configurable
- sinks - I could imagine some situation when you want to specify behaviour of sink during deployment
We could introduce a sealed trait for that types of components, I mean `NonBuiltInComponent`. `CustomComponent` would be better but it is already reserved for `CustomNode`.

Reply:
I fixed others, but I have to find out how to compile all nodes at once

Reply:
You don't need to compile all, built-in components can't be ommitted